### PR TITLE
Fix for "Azure DevOps Pipelines fails git with an HTTP 400 error"

### PR DESCRIPTION
**Pull request by tatyana-kostromskaya**
Sometimes checkout task may fail with the following message:

```
fatal: unable to access 'https://.../': The requested URL returned error: 400
Git fetch failed with exit code: 128
```

The git configuration change caused the duplicate authentication header which in turn resulted in the Http 400 error across pipeline.
I was able to reproduce this issue on my local agent by setting "http.extraheader" key in local git config with invalid value. Then I ran the build and got the same error message.

Add ability to get values of pre-existing `extraheader` configs to be able to clean up them from git config file.
Link to the ICM: https://portal.microsofticm.com/imp/v3/incidents/details/275689076/home

Tested manually on self-hosted agent with additional logic to add manually several extraheaders.

### CODE REVIEW

File: src/Agent.Plugins/GitCliManager.cs [hunk 1]
```diff
@@ -441,6 +441,22 @@ public async Task<bool> GitConfigExist(AgentTaskPluginExecutionContext context,
             return exitcode == 0;
         }
 
+        // git config --get-all <key>
+        public async Task<bool> GitConfigExist(AgentTaskPluginExecutionContext context, string repositoryPath, string configKey, IList<string> existingConfigValues)
```
Comment (anatolybolshakov) [inline src/Agent.Plugins/GitCliManager.cs:445]: Please add summary comments to this new method

**Comment by anatolybolshakov** [comment]
@tatyana-kostromskaya let's perform some buddy testing as well here

File: src/Agent.Plugins/GitSourceProvider.cs [hunk 2]
```diff
@@ -657,11 +657,17 @@ public string GenerateAuthHeader(AgentTaskPluginExecutionContext executionContex
                 executionContext.Debug($"Remove http.{repositoryUrl.AbsoluteUri}.extraheader setting from git config.");
                 await RemoveGitConfig(executionContext, gitCommandManager, targetPath, $"http.{repositoryUrl.AbsoluteUri}.extraheader", string.Empty);
             }
-            if (await gitCommandManager.GitConfigExist(executionContext, targetPath, $"http.extraheader"))
+
+            var existingExtraheaders = new List<string>();
+            if (await gitCommandManager.GitConfigExist(executionContext, targetPath, $"http.extraheader", existingExtraheaders))
```
Comment (anatolybolshakov) [inline src/Agent.Plugins/GitSourceProvider.cs:662]: Why do we need to pass existingExtraheaders here if it's always empty?

**Comment by anatolybolshakov** [comment]
Please describe the problem in more details in description

File: src/Agent.Plugins/GitSourceProvider.cs (hunk 2 above)
Comment (anatolybolshakov) [reply src/Agent.Plugins/GitSourceProvider.cs:662]: See it now - existingExtraheaders is the output array (which looks unobvious here).
Please describe it in method's comments also

**Comment by tatyana-kostromskaya** [comment]
> Please describe the problem in more details in description

I have added more details

File: src/Agent.Plugins/GitCliManager.cs [hunk 3]
```diff
@@ -441,6 +441,29 @@ public async Task<bool> GitConfigExist(AgentTaskPluginExecutionContext context,
...
+        {
+            // git config --get-all {configKey} will return 0 and print the value if the config exist.
+            context.Debug($"Checking git config {configKey} exist or not");
+
+            // ignore any outputs by redirect them into a string list, since the output might contains secrets.
+            if (existingConfigValues == null)
+            {
+                existingConfigValues = new List<string>();
+            }
```
[the remark this answers is not in the document]
Comment (anatolybolshakov) [reply src/Agent.Plugins/GitCliManager.cs:461]: @KonstantinTyukalov this is only for existingConfigValues == null case, for not empty value case there could be some values in list

**Review by KonstantinTyukalov — APPROVED** [review]
LGTM, thanks! Please check unresolved comments

**Review by anatolybolshakov — APPROVED** [review]

**Review by DenisRumyantsev — APPROVED** [review]

**Pull request merged**